RecoParticleFlow/PFClusterProducer/plugins/PFBadHcalPseudoClusterProducer.cc

```diff
@@ -0,0 +1,176 @@
+        if (minD.second == HcalEmpty || minD.first > id.depth()) {
+            minD.first  = id.depth();
+            minD.second = id.subdet();
+        }
+    }
+
+    const float dummyEnergy = 1e-5; // non-zero, but small (even if it shouldn't ever be used)
+    for (const auto & rec : bads) {
+        int ieta = rec.first.first, iphi = rec.first.second, nbad = rec.second, ngood = good[rec.first];
+        auto minDepth = minDepths[rec.first];   
+        bool barrel = minDepth.second == HcalBarrel;
+        HcalDetId id(minDepth.second, ieta, iphi, minDepth.first);
+        bool isBad = (nbad > 0 && nbad >= ngood);
+        if (debug_) std::cout << "At ieta " << id.ieta() << ", iphi " << id.iphi() << " I have " << nbad << " bad depths, " << ngood << " good depths. First depth is in " << (barrel ? "HB" : "HE") << " depth " << minDepth.first <<"; " << (isBad ? " MARK BAD": " ignore") << std::endl;
+        if (!isBad) continue;
+        PFLayer::Layer layer = (barrel ? PFLayer::HCAL_BARREL1 : PFLayer::HCAL_ENDCAP);
+        // make a PF RecHit
+        std::shared_ptr<const CaloCellGeometry> thisCell = (barrel ? hbGeom : heGeom)->getGeometry(id);
+        const GlobalPoint & pos = thisCell->getPosition();
+        badAreasRH_.emplace_back( thisCell, id(), layer, dummyEnergy );
+        // make a PF cluster (but can't add the rechit, as for that I need an edm::Ref)
+        badAreasC_.emplace_back( layer, dummyEnergy, pos.x(), pos.y(), pos.z() );
+        badAreasC_.back().setFlags(reco::CaloCluster::badHcalMarker);
+    }
+
+    cacheId_quality_ = iSetup.get<HcalChannelQualityRcd>().cacheIdentifier();
+    cacheId_geom_ = iSetup.get<CaloGeometryRecord>().cacheIdentifier();
+}
+
+// ------------ method called to produce the data  ------------
+    void
+PFBadHcalPseudoClusterProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+    if (enabled_) {
+        if (cacheId_quality_ != iSetup.get<HcalChannelQualityRcd>().cacheIdentifier() ||
+               cacheId_geom_ != iSetup.get<CaloGeometryRecord>().cacheIdentifier()) {
+            init(iSetup);
+        }
+    }
+
+
+    auto outRH = std::make_unique<std::vector<reco::PFRecHit>>(badAreasRH_);
+    auto rhHandle = iEvent.put(std::move(outRH), "hits");
+
+    auto outC = std::make_unique<std::vector<reco::PFCluster>>(badAreasC_);
+
+    // now we go set references
+    for (unsigned int i = 0, n = rhHandle->size(); i < n; ++i) {
+        (*outC)[i].addRecHitFraction( reco::PFRecHitFraction(reco::PFRecHitRef(rhHandle,i), 1.0) );
+    }
+
+    iEvent.put(std::move(outC));
+}
+
+void PFBadHcalPseudoClusterProducer::fillDescriptions(edm::ConfigurationDescriptions & descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<bool>("enable", true);
+    desc.add<int32_t>("thresholdHB",  1)->setComment("minimum required number of bad depths in HB (not including HO)");
+    desc.add<int32_t>("thresholdHE",  2)->setComment("minimum required number of bad depths in HE");
```

Comment:
these two are not needed anymore.

Also, it's unclear to me if ```enable``` should be true by default. It is pseudo-set to true later by ```PF_badHcalMitigation```, which seems redundant (or suspicious).

Reply:
Now the default of the code is `true`, the default in the configuration cff is `false`, and is turned back to `true` by the era modifier. 
I could change the default of the code to be false, but eventually the long term goal is to  have this turned on by default, so I think it makes more sense to have the code default to true.